### PR TITLE
Potentially cheaper array based LRU cache

### PR DESCRIPTION
**Pull request by Scooletz**
This is a speculative PR proposing changes for the implementation of the LRU cache.

Previous version was based on `LinkedList<T>`,  `LinkedListNode<T>` and `LruCacheItem`. It had following properties:

- each of this types is a class, therefore it allocates an object header
- `LinkedListNode<T>` contains three references: `previous`, `next` and the `list`. The value is stored in the `item`
- `LruCacheItem` contains the key (needed for `.Replace` operation) and the actual value.

When profiling a runner, I found large amounts of objects allocated for the purpose of `LinkedListNode<T>` and `LruCacheItem`.

This PR changes the approach to use an **array of structs** and implementing a linked list on to of it. It removes some memory overhead as a single array item consist of:

- `prev` and `next` implemented as 32bit ints, no references needed
- `TValue` and `TKey` like in the previously existing `LruCacheItem`
- no reference to the list in a struct (not needed).

**This is an exploratory work and requires more testing and benchmarking**. Initial memory profiling looked possibly promising (I'm trying to use soft language to do not overpromise here), showing removal of a lot of objects' litter.

#### LinkedList based approach

|  Item | Description  | 32-bit | 64-bit  |
|---|---|---|---|
| LinkedListNode | class overhead (object header)  |  8  | 16  |
| LinkedListNode.prev |  an object reference  | 4  | 8  |
| LinkedListNode.next  |  an object reference  | 4  | 8  |
| LinkedListNode.list  | an object reference  | 4  | 8  |
| LruCacheItem  | class overhead (object header)  | 8  | 16  |
|  **Sum** |   | 24  | 48  |

#### Array of structs based approach

|  Item | Description  | 32-bit | 64-bit  |
|---|---|---|---|
| Cell.prev | an int32  |  4  | 4  |
| Cell.next | an int32  |  4  | 4  |
|  **Sum** |   | 8  | 8  |

#### Difference

|  Approach | 32-bit | 64-bit  |
|---|---|---|
| LinkedList |  24  | 48  |
| Array of structs |  8 | 8 |
| **Gain** |  16 | 40 |

This just a size estimation for a single item. It does not take into consideration GC overhead for keeping a lot of objects alive.

### Benchmarks

I used and modified `LruCacheAddAtCapacityBenchmarks`

#### Before

|         Method |     Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |---------:|----------:|----------:|------:|------:|------:|----------:|
| WithRecreation | 4.736 ms | 0.0930 ms | 0.0955 ms |     - |     - |     - |   1.92 KB |
|      WithClear | 4.661 ms | 0.1011 ms | 0.0946 ms |     - |     - |     - |   1.26 KB |

#### After

|         Method |     Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |---------:|----------:|----------:|------:|------:|------:|----------:|
| WithRecreation | 3.625 ms | 0.0672 ms | 0.0628 ms |     - |     - |     - |     917 B |
|      WithClear | 3.573 ms | 0.0620 ms | 0.0580 ms |     - |     - |     - |     430 B |

### CODE REVIEW

**Comment by Scooletz** [comment]
@tkstanczak I flushed some estimations for sizes that we can discuss taking into consideration count of items and caching strategies (where it's cleared etc.)

**Comment by tkstanczak** [comment]
@Scooletz I am ok to merge this but cannot merge a draft

**Comment by Scooletz** [comment]
@tkstanczak added modified benchmark and a benchmark result

File: src/Nethermind/Nethermind.Benchmark/Core/LruCacheAddAtCapacityBenchmarks.cs [hunk 1]
```diff
@@ -20,33 +20,39 @@
 namespace Nethermind.Benchmarks.Core
 {
     [MemoryDiagnoser]
-    [SimpleJob]
     public class LruCacheAddAtCapacityBenchmarks
     {
+        const int Capacity = 16;
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Benchmark/Core/LruCacheAddAtCapacityBenchmarks.cs:25]: strange naming - maybe default or initial

Comment (tkstanczak) [reply src/Nethermind/Nethermind.Benchmark/Core/LruCacheAddAtCapacityBenchmarks.cs:25]: aaaaah sorry, this is benchmark

Comment (tkstanczak) [reply src/Nethermind/Nethermind.Benchmark/Core/LruCacheAddAtCapacityBenchmarks.cs:25]: ignore

**Review by tkstanczak — APPROVED** [review]

**Pull request merged**